File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(strings, predicate)
  for _, value in pairs(strings) do
    if predicate(value) == true
    then
      return string.lower(value)
    end

  end

  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.create(function ()
    local exp = 0
    while true do
      local num = base ^ exp
      if num > limit then
        break
      else
        coroutine.yield(num)
      end
        exp = exp + 1
    end
  end)
end

-- Write your say function here
function say(str)
  print("\n")
  
  local words = {}

  if str ~= nil then
    table.insert(words, str)  
  end

  if str == nil and #words == 0 then
    return ""
  end

  function addWord(newStr)
    if newStr == nil then
      return table.concat(words, " ")
    else
      table.insert(words, newStr)
      return addWord
    end
  end

  return addWord
end

-- Write your line count function here
function meaningful_line_count(filename)
  local file = io.open(filename, 'r')
  local line_count = 0

  if file then
    for line in file:lines() do
      local str = line:match("^%s*(%S)")
      if not(str == nil or str:sub(1,1) == '#') then
        line_count = line_count + 1
      end
    end
    file:close()
    
  else
    error("No such file")
  end
    
  return line_count
end


-- Write your Quaternion table here
Quaternion = {}

quatmeta = {
  __index = {
    coefficients = function(self)
      return {self.a, self.b, self.c, self.d}
    end,

    conjugate = function(self)
      print(Quaternion.new(self.a, -self.b, -self.c, -self.d))
      return Quaternion.new(self.a, -self.b, -self.c, -self.d)
    end
  },
  __add = function(self, other)
    return Quaternion.new(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
  end,
  __mul = function(self, other)
    return Quaternion.new(self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
                          self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
                          self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
                          self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a)
  end,
  __tostring = function(self)
    local final_string = ""

    -- Edge Case where everything is 0
    if self.a == 0 and self.b == 0 and self.c == 0 and self.d == 0 then
        return "0"
    end

    -- First coefficient
    if self.a ~= 0 then
        final_string = final_string .. tostring(self.a)
    end

    -- Second coefficient (i)
    if self.b ~= 0 then
        if final_string == "" and math.abs(self.b) ~= 1 then
            final_string = final_string .. tostring(self.b) .. 'i'
        elseif final_string == "" and self.b == 1 then
            final_string = final_string .. 'i'
        elseif final_string == "" and self.b == -1 then
            final_string = final_string .. '-i'
        else
            if self.b > 0 then
                final_string = final_string .. "+"
            else
                final_string = final_string .. "-"
            end

            if math.abs(self.b) ~= 1 then
                final_string = final_string .. tostring(math.abs(self.b)) .. 'i'
            else
                final_string = final_string .. 'i'
            end
        end
    end

    -- Third coefficient (j)
    if self.c ~= 0 then
        if final_string == "" and math.abs(self.c) ~= 1 then
            final_string = final_string .. tostring(self.c) .. 'j'
        elseif final_string == "" and self.c == 1 then
            final_string = final_string .. 'j'
        elseif final_string == "" and self.c == -1 then
            final_string = final_string .. '-j'
        else
            if self.c > 0 then
                final_string = final_string .. "+"
            else
                final_string = final_string .. "-"
            end

            if math.abs(self.c) ~= 1 then
                final_string = final_string .. tostring(math.abs(self.c)) .. 'j'
            else
                final_string = final_string .. 'j'
            end
        end
    end

    -- Fourth coefficient (k)
    if self.d ~= 0 then
        if final_string == "" and math.abs(self.d) ~= 1 then
            final_string = final_string .. tostring(self.d) .. 'k'
        elseif final_string == "" and self.d == 1 then
            final_string = final_string .. 'k'
        elseif final_string == "" and self.d == -1 then
            final_string = final_string .. '-k'
        else
            if self.d > 0 then
                final_string = final_string .. "+"
            else
                final_string = final_string .. "-"
            end

            if math.abs(self.d) ~= 1 then
                final_string = final_string .. tostring(math.abs(self.d)) .. 'k'
            else
                final_string = final_string .. 'k'
            end
        end
    end

    return final_string
  end,

  __eq = function(self, other)
    return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d
  end
}

Quaternion.new = function(a, b, c, d)
  return setmetatable({a = a, b = b, c = c, d = d}, quatmeta)
end

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import *
import os


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(strings: list[str], function: Callable[[str], bool]) -> str:
    for string in strings:
        if function(string) == True:
            return string.lower()
    
    return None

# Write your powers generator here
def powers_generator(base: int, limit: int) -> Generator:
    exp: int = 0
    
    while True:
        num = pow(base, exp)
        if (num > limit):
            break
        else:
            exp += 1
            yield num





# Write your say function here
def say(string: Optional[str] = None) -> Callable[..., str]:
    list_of_words: list[str] = []

    if string is None and len(list_of_words) == 0:
        return ""

    def inner(next_string: Optional[str] = None) -> str:
        if next_string is None:
            return " ".join(list_of_words)
        else:
            list_of_words.append(next_string)
            return inner

    if string is not None:
        list_of_words.append(string)

    return inner

# Write your line count function here

def meaningful_line_count(filename: str) -> int:
    line_count: int = 0
    try:
       with open(filename, 'r') as file:
           for line in file:
               if (line.strip() != "" and line.strip()[0] != '#'):
                   line_count += 1
               
    except:
        raise FileNotFoundError('No such file')
    
    return line_count

# Write your Quaternion class here
@dataclass(frozen = True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float
    
    def __str__(self) -> str:
        s = ""
        co_eff = self.coefficients  
        pairs = zip(co_eff, ["", "i", "j", "k"])

        for num, suffix in pairs:
            if num == 0:
                continue
            
            #Condition when string is empty
            if s == '':
                if suffix == "" or abs(num) != 1:  
                    s += str(num)
                else:
                    s += "-" if num == -1 else ""  
            else:
                # If string is NOT empty, checks what sign to add
                if num > 0:
                    s += "+"
                else:
                    s += "-"
                
                # Adds number
                if suffix == "" or abs(num) != 1:
                    s += str(abs(num))
            
            # Adds suffix
            s += suffix
        
        if s == "":
            return "0"
        
        return s


    def __add__(self, other) -> object:
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
    
    def __mul__(self, other: object) -> object:
        return Quaternion(self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
                          self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
                          self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
                          self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a)
    
    def __eq__(self, value: object) -> bool:
        return self.a == value.a and self.b == value.b and self.c == value.c and self.d == value.d
    
    @property
    def conjugate(self) -> object:
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    @property
    def coefficients(self) -> object:
        return (self.a, self.b, self.c, self.d)
    
